README
Qt Components for BB10
===============================

This branch (bb10) contains the MeeGo/Harmattan Qt Components modified to work on the BlackBerry 10 OS.

Modifications done
==================
- main window size is automatically detected as runtime ("normal" Qt Components just have a hardcoded value)
- the default orientation was changed to portrait
- orientation fixes to work correctly with the QBB_USE_OPENGL environmental variable
- removed dependency on QSystemInfo when building with qt-mobility support, as QSystemInfo is not part of the BB10 Qt build
- fixed screen rotation to landscape every time the application is minimized
- high-DPI adaptation for relevant components --> this is not 100% done & patches are welcome ! :)
-- basically, most of the hardcoded values were multiplied by 2, which seems to work quite fine on the high-DPI screen of the Dev-Alpha

Where to make changes
=====================
For you changes to make effect, edit the files present in the src/fremantle/cssu folder.
Changing the other files will not influence the build with the below provided commands.

Preparing fo the build
======================
Firs of all install the BlackBerry 10 native SDK.
Then prepare you shell for cross-compiling with the BB10 NDK:
source ~/bbndk/bbdnk-env.sh

Configuring
===========
Use the following command to configure for build:
./configure -fremantle -no-maliit -make cssu -force-local-theme -mobility -no-xrandr -no-xdamage -nomake examples

Compiling
=========
Just start make to compile it:
make -j5

Result
======
The resulting compiled components are in the imports/ folder.
You can drop the "org" folder as that contains components targeted for Maemo 5 Fremantle,
that are not needed by the ordinary MeeGo components.

Recompiling
===========
First clean the directory:
make clean
Then remove the imports folder with the result & proceed with configure & make as usual.

Using the components
====================
To use these components, use the following import in your QML:
import com.nokia.meego 1.0
You also need to pack the contents of the "imports" folder together with your application and then point the QML_IMPORT_PATH to it on application startup.
Or you can feel the QDeclarativeEngine to add it when instantiation the QDeclarativeView.
Also don't forget to set the QBB_USE_OPENGL environmental variable to "1", as it it would not use OpenGL and be horrendously slow.
The automatic orientation code also expects OpenGL to be set, as it fixes a few glitches introduces by using it and might not work correctly otherwise.
You can include the following code in your bar, which makes sure QBB_USE_OPENGL is properly set at startup:
<env var="QBB_USE_OPENGL" value="1"/>

Original Qt Components README
=============================
This source tree contains the Qt Components project.

Building
========
configure (for compiling specific component sets add '-meego' or '-symbian')
make
make install

The project is divided into:

- The module Qt.labs.components: _for now_ contains general purpose helpers for QML.

and also (for the time being) two QML modules for component sets implemented using those components

- com.meego: components from MeeGo Touch UI library,
- com.nokia.symbian: set of UI components for Symbian.

The lib will install in QTDIR/lib, while the modules will go to QTDIR/imports.

Build dependencies
==================
- Qt 4.7

Optional dependencies
=====================
- MeeGo Touch (for some features of the the com.meego module)
  -- On desktop, download and install: libmeegotouch & meegotouch-theme from http://meego.gitorious.org/meegotouch
  -- NOTE: meegotouch is already available if you are compiling inside scratchbox with a maemo6 target setup

Coding Style
==================
Follow the respective Qt coding style for C++ and QML files and the
JavaScript coding style for JS only files.

Coding Styles
- Qt C++: http://qt.gitorious.org/qt/pages/QtCodingStyle
- QML: http://doc.qt.nokia.com/4.7/qml-coding-conventions.html
- JS: https://developer.mozilla.org/En/Developer_Guide/Coding_Style

Running on Desktop
==================
- Pass -graphicssystem native
- On mac, running tests require prefixing the path to the libqtcomponentsplugin.dylib: 
   "DYLD_LIBRARY_PATH=/Developer/depot/qt/imports/Qt/labs/components/ make check"

Building the Documentation
==================

To generate the Symbian components documentation:
configure -symbian
make docs

This will generate HTML files in the doc/html directory. Also, a qch file will
be generated in the doc/qch directory. This qch file can be loaded in
Qt Assistant or Qt Creator.

Build instructions for MeeGoTouch (quickstart)
==============================================
(After trying to install things in a different prefix,
 I gave up since I couldn't get meegotouch to find its theme.)

You will need these projects:
git://gitorious.org/meegotouch/meegotouch-theme.git
git://gitorious.org/meegotouch/libmeegotouch.git
git@gitorious.org:qt-components/qt-components.git

1: meegotouch-theme: qmake PREFIX=/usr/local/ && sudo make install

2: libmeegotouch:
   ./configure --prefix=/usr
 Make sure that you configure with dbus. If it doesn't pick up qt's dbus:
   export PKG_CONFIG_PATH=$QTDIR/lib/pkgconfig
   make install

3: qt-components: ./configure -meego && make install
     (installs into your current binary qt path)
     or use dpkg-buildpackage to make a .deb package


debian/changelog
qt-components (1.0~git20110712-1+0m5) fremantle; urgency=low

  * Initial port to fremantle

 -- Carlos Martin <cmartin@liberalia.net>  Tue, 12 July 2011 19:55:00 +0300

qt-components (1.0~git20110712-1) maverick; urgency=low

  * Fixes: NB#270426 - coreweb: /usr/bin/FacebookQML "Fails at Daemon::readFromBoosterSocket" 

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Tue, 12 July 2011 13:55:00 +0300

qt-components (1.0~git20110630-1) maverick; urgency=low

  * Fixes: NB#268578 - Language change not reflected when QML app is launched from invoker

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Thu, 30 June 2011 15:10:00 +0300

qt-components (1.0~git20110628-1) maverick; urgency=low

  * Fixes: NB#268223 Paste function fails work if the text box has been pasted once after copying text from Notes app 

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Tue, 28 June 2011 13:40:00 +0300

qt-components (1.0~git20110621-1) maverick; urgency=low

  * Fixes: NB#267339 It's not possible to move the cursor to the last positions of a line with the magnifier
  * Fixes: NB#267240 LEGAL BUG Update qt-components sources with latest license

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Tue, 21 June 2011 14:40:00 +0300

qt-components (1.0~git20110620-3) maverick; urgency=low

  * Fixes: NB#266959 - Tabs activate on tap down

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Mon, 20 June 2011 15:35:00 +0300

qt-components (1.0~git20110620-2) maverick; urgency=low

  * Fixes: NB#267037 Tap down effect on the tab is missing

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Mon, 20 June 2011 14:17:00 +0300

qt-components (1.0~git20110620-1) maverick; urgency=low

  * Fixes: NB#265642 - TextField not repositioned when entering input after panning the screen
  * Fixes: NB#264962 - Wrong textfield is shown when repositioning screen 

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Mon, 20 June 2011 13:54:00 +0300

qt-components (1.0~git20110616-1) maverick; urgency=low

  * Fixes: NB#265212 VKB orientation is set to landscape even when page orientation is locked to portrait

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Thu, 16 June 2011 08:28:00 +0300

qt-components (1.0~git20110615-1) maverick; urgency=low

  * Fixes: NB#265115 Facebook status bar dissappears on new status update 
  * Fixes: NB#265658 MRemoteThemeDaemonClient creates unnecessary pixmaps

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Wed, 15 June 2011 12:04:00 +0300

qt-components (1.0~git20110614-1) maverick; urgency=low

  * Fixes: NB#264944 typing in pre-edit doesn't reposition the textfield
  * Fixes: NB#264577 Word tracker is wrongly placed after screen orientation change 

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Tue, 14 June 2011 14:42:00 +0300

qt-components (1.0~git20110613-1) maverick; urgency=low

  * Fixes: NB#263808 Editbubble does not go over statusbar on top of the page
  * Fixes: NB#263802 Editbubble is closed on press
  * Fixes: NB#262911 Editbubble doesn't appear after deactivating a word by tapping to whitespace

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Mon, 13 June 2011 10:59:00 +0300

qt-components (1.0~git20110610-1) maverick; urgency=low

  * Fixes: Please see 0613-1

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Fri, 10 June 2011 13:03:00 +0300

qt-components (1.0~git20110608-1) maverick; urgency=low

  * Fixes: NB#262006 Severe splash screen transition flickering problem in new QML calculator
  * Fixes: NB#260684 Inverted SheetStyle not respected for Sheet 

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Wed, 08 June 2011 11:40:00 +0300

qt-components (1.0~git20110531-2) maverick; urgency=low

  * Fixes a failing test case

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Tue, 31 May 2011 13:27:00 +0300

qt-components (1.0~git20110531-1) maverick; urgency=low

  * Fixes: NB#259158 Qt components text entries clear clipboard even though paste failed
  * Fixes: NB#260191 Please check sheet transition parameter

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Tue, 31 May 2011 12:30:00 +0300

qt-components (1.0~git20110527-1) maverick; urgency=low

  * Fixes: NB#258670 - VKB is not displayed properly when QML component gallery opened..

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Fri, 27 May 2011 15:22:00 +0300

qt-components (1.0~git20110525-1) maverick; urgency=low

  * Fixes: NB#254399 Magnifying lens hidden in Facebook
  * Fixes: NB#257472 Single item in menu doesn't draw the rounder corners to bottom

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Wed, 25 May 2011 14:16:00 +0300

qt-components (1.0~git20110524-1) maverick; urgency=low

  * Fixes: NB#256227 Text entry paste does not clear clipboard
  * Fixes: NB#256483 Hard to place the text area cursor position 0 in QCG compared to libmeegotouch
  * Fixes: NB#257249 Textured (not smooth) icons in TabButtons
  * Fixes: NB#258863 Virtual keyboard does not disappear when TextField loses focus
  * Fixes: NB#257880 Edit bubble hides selected text

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Tue, 24 May 2011 13:16:00 +0300

qt-components (1.0~git20110519-1) maverick; urgency=low

  * Fixes: NB#249316 Edit bubble is not repositioned when view is scrolled
  * Fixes: NB#255452 Word tracker doesn't follow view panning
  * Fixes: NB#257570 In inverted landscape status bar is render for portrait mod

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Thu, 19 May 2011 16:13:00 +0300

qt-components (1.0~git20110518-1) maverick; urgency=low

  * Fixes: NB#253997 Time Picker looping in AM/PM column causes erratic behavior
  * Fixes: NB#255634 Default width and font-size of toolbar buttons are not right
  * Fixes: NB#256728 Binding ToolBar opacity fail

 -- Joonas Reynders <ext-joonas.n.reynders@nokia.com>  Wed, 18 May 2011 12:48:00 +0300

qt-components (1.0~git20110517-1) maverick; urgency=low


src/fremantle/plugin.cpp
/****************************************************************************
**
** Copyright (C) 2011 Nokia Corporation and/or its subsidiary(-ies).
** All rights reserved.
** Contact: Nokia Corporation (qt-info@nokia.com)
**
** This file is part of the Qt Components project.
**
** $QT_BEGIN_LICENSE:BSD$
** You may use this file under the terms of the BSD license as follows:
**
** "Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are
** met:
**   * Redistributions of source code must retain the above copyright
**     notice, this list of conditions and the following disclaimer.
**   * Redistributions in binary form must reproduce the above copyright
**     notice, this list of conditions and the following disclaimer in
**     the documentation and/or other materials provided with the
**     distribution.
**   * Neither the name of Nokia Corporation and its Subsidiary(-ies) nor
**     the names of its contributors may be used to endorse or promote
**     products derived from this software without specific prior written
**     permission.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
** "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
** LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
** A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
** OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
** LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
** $QT_END_LICENSE$
**
****************************************************************************/

//Add here type headers
#include "mdeclarativemousefilter.h"
#include "mdeclarativeimageprovider.h"
#include "mdeclarativeimobserver.h"
#include "mdeclarativeimplicitsizeitem.h"
#include "mdeclarativeinputcontext.h"
#include "mdeclarativescreen.h"
#include "minversemousearea.h"
#include "mscrolldecoratorsizer.h"
#include "msnapshot.h"
#include "mtexttranslator.h"
#include "mthemeplugin.h"
#include "mwindowstate.h"
#include "plugin.h"
#include "sdeclarative.h"

#include <QDeclarativePropertyMap>
#include <QFont>

#ifdef HAVE_SHADER
# include "shadereffectitem/shadereffectitem.h"
# include "shadereffectitem/shadereffectsource.h"
#else
# include "shadereffectitem/shadereffectitemnull.h"
# include "shadereffectitem/shadereffectsourcenull.h"
#endif

#include <QApplication>
#include <QtDeclarative>

static const int VERSION_MAJOR = 1;
static const int VERSION_MINOR = 0;

void FremantlePlugin::initializeEngine(QDeclarativeEngine *engine, const char *uri) {
        
        QDeclarativeExtensionPlugin::initializeEngine(engine, uri);
        QDeclarativeContext *context = engine->rootContext();

        // Intentionally override possible older version of the plugin.
        context->setProperty("fremantleComponentsVersionMajor", VERSION_MAJOR);
        context->setProperty("fremantleComponentsVersionMinor", VERSION_MINOR);

        // SetUp Theme image provider
        engine->addImageProvider(QLatin1String("theme"), new MDeclarativeImageProvider);

        // Register global screen instance
        context->setContextProperty("screen", MDeclarativeScreen::instance());
        qmlRegisterUncreatableType<MDeclarativeScreen>(uri, 1, 0, "Screen", "");

        // Theme support
        context->setContextProperty("theme", new MThemePlugin);
        qmlRegisterUncreatableType<MThemePlugin>(uri, 1, 0, "Theme", "");

        // Register global ImputContext support (Keyboard Stuff)
        context->setContextProperty("inputContext", new MDeclarativeInputContext);
        qmlRegisterUncreatableType<MDeclarativeInputContext>(uri, 1, 0, "InputContext", "");

        SDeclarative *declarative = new SDeclarative(context);
        context->setContextProperty("maemo", declarative);

        context->setContextProperty("platformWindow", MWindowState::instance());
        qmlRegisterUncreatableType<MWindowState>(uri, 1, 0, "WindowState", "");

        context->setContextProperty("textTranslator", new MTextTranslator);
        qmlRegisterUncreatableType<MTextTranslator>(uri, 1, 0, "TextTranslator", "");

        // Disable cursor blinking + make double tapping work the way it is done in lmt.
        QApplication *app = qobject_cast<QApplication*>(QApplication::instance());
        if (app) {
            app->setCursorFlashTime(0);
            app->setDoubleClickInterval(MEEGOTOUCH_DOUBLETAP_INTERVAL);
        } 

        // Set style constants
        context->setContextProperty("UiConstants", uiConstants());
}

void FremantlePlugin::registerTypes(const char *uri) {
        // Add here custom types
        qmlRegisterType<MSnapshot>(uri, 1, 0, "Snapshot");
        qmlRegisterUncreatableType<MWindowState>(uri, 1, 0, "WindowState","");
        qmlRegisterUncreatableType<MDeclarativeScreen>(uri, 1, 0, "Screen", "");
        qmlRegisterUncreatableType<MDialogStatus>(uri, 1, 0, "DialogStatus", "");
        qmlRegisterUncreatableType<SPageOrientation>(uri, 1, 0, "PageOrientation", "");
        qmlRegisterUncreatableType<SPageStatus>(uri, 1, 0, "PageStatus", "");
        qmlRegisterUncreatableType<MToolBarVisibility>(uri, 1, 0, "ToolBarVisibility", "");

        // Custom primitives
        qmlRegisterType<MDeclarativeImplicitSizeItem>(uri, 1, 0, "ImplicitSizeItem");
        qmlRegisterType<MDeclarativeFocusScope>(uri, 1, 0, "FocusScope");
        qmlRegisterType<MInverseMouseArea>(uri, 1, 0, "InverseMouseArea");
        qmlRegisterType<MDeclarativeMouseFilter>(uri, 1, 0, "MouseFilter");
        qmlRegisterType<MDeclarativeMouseEvent>(uri, 1, 0, "MMouseEvent");
        qmlRegisterType<MDeclarativeIMObserver>(uri, 1, 0, "InputMethodObserver");
        qmlRegisterType<MScrollDecoratorSizer>(uri, 1, 0, "ScrollDecoratorSizerCPP");

        // shader effect item (to be removed when supported in QML)
#ifdef HAVE_SHADER
        qmlRegisterType<ShaderEffectItem>(uri, 1, 0, "ShaderEffectItem");
        qmlRegisterType<ShaderEffectSource>(uri, 1, 0, "ShaderEffectSource");
#else
        qmlRegisterType<ShaderEffectItemNull>(uri, 1, 0, "ShaderEffectItem");
        qmlRegisterType<ShaderEffectSourceNull>(uri, 1, 0, "ShaderEffectSource");
#endif
}

QDeclarativePropertyMap *FremantlePlugin::uiConstants() {
        QDeclarativePropertyMap *uiConstantsData = new QDeclarativePropertyMap();
        uiConstantsData->insert("DefaultMargin", QVariant(16));
        uiConstantsData->insert("ButtonSpacing", QVariant(6));
        uiConstantsData->insert("HeaderDefaultHeightPortrait", QVariant(72));
        uiConstantsData->insert("HeaderDefaultHeightLandscape", QVariant(46));
        uiConstantsData->insert("HeaderDefaultTopSpacingPortrait", QVariant(20));
        uiConstantsData->insert("HeaderDefaultBottomSpacingPortrait", QVariant(20));
        uiConstantsData->insert("HeaderDefaultTopSpacingLandscape", QVariant(16));
        uiConstantsData->insert("HeaderDefaultBottomSpacingLandscape", QVariant(14));
        uiConstantsData->insert("ListItemHeightSmall", QVariant(64));
        uiConstantsData->insert("ListItemHeightDefault", QVariant(80));

        QFont bodyTextFont;
        bodyTextFont.setFamily("Nokia Pure Text Light");
        bodyTextFont.setPixelSize(24);
        uiConstantsData->insert("BodyTextFont", QVariant(bodyTextFont));

        QFont headerFont;
        headerFont.setFamily("Nokia Pure Text Light");
        headerFont.setPixelSize(32);
        uiConstantsData->insert("HeaderFont", QVariant(headerFont));

        QFont groupHeaderFont;
        groupHeaderFont.setFamily("Nokia Pure Text");
        groupHeaderFont.setPixelSize(18);
        groupHeaderFont.setBold(true);
        uiConstantsData->insert("GroupHeaderFont", QVariant(groupHeaderFont));

        QFont titleFont;
        titleFont.setFamily("Nokia Pure Text");
        titleFont.setPixelSize(26);
        titleFont.setBold(true);
        uiConstantsData->insert("TitleFont", QVariant(titleFont));

        QFont smallTitleFont;
        smallTitleFont.setFamily("Nokia Pure Text");
        smallTitleFont.setPixelSize(24);
        smallTitleFont.setBold(true);
        uiConstantsData->insert("SmallTitleFont", QVariant(titleFont));

        QFont fieldLabelFont;
        fieldLabelFont.setFamily("Nokia Pure Text Light");
        fieldLabelFont.setPixelSize(22);
        uiConstantsData->insert("FieldLabelFont", QVariant(fieldLabelFont));
        uiConstantsData->insert("FieldLabelColor", QVariant(QColor("#505050")));

        QFont subTitleFont;
        subTitleFont.setFamily("Nokia Pure Text Light");
        subTitleFont.setPixelSize(22);
        uiConstantsData->insert("SubtitleFont", QVariant(subTitleFont));

        return uiConstantsData;
}

#include "moc_plugin.cpp"

Q_EXPORT_PLUGIN2(fremantleplugin, FremantlePlugin);
